src/common/SubProcess.h

```diff
@@ -236,253 +136,10 @@ class fd_buf : public std::streambuf {
-  // to ignore it.
-  if (signal(SIGTERM, SIG_DFL) == SIG_ERR) {
-    std::cerr << cmd << ": signal failed: " << cpp_strerror(errno) << "\n";
-    goto fail_exit;
-  }
-  // Because SIGCHLD is ignored by default, setup dummy handler for it,
-  // so we can mask it.
-  if (signal(SIGCHLD, dummy_sighandler) == SIG_ERR) {
-    std::cerr << cmd << ": signal failed: " << cpp_strerror(errno) << "\n";
-    goto fail_exit;
-  }
-  // Setup timeout handler.
-  if (signal(SIGALRM, timeout_sighandler) == SIG_ERR) {
-    std::cerr << cmd << ": signal failed: " << cpp_strerror(errno) << "\n";
-    goto fail_exit;
-  }
-  // Block interesting signals.
-  sigemptyset(&mask);
-  sigaddset(&mask, SIGINT);
-  sigaddset(&mask, SIGTERM);
-  sigaddset(&mask, SIGCHLD);
-  sigaddset(&mask, SIGALRM);
-  if (sigprocmask(SIG_SETMASK, &mask, &oldmask) == -1) {
-    std::cerr << cmd << ": sigprocmask failed: " << cpp_strerror(errno) << "\n";
-    goto fail_exit;
-  }
-
-  pid = fork();
-
-  if (pid == -1) {
-    std::cerr << cmd << ": fork failed: " << cpp_strerror(errno) << "\n";
-    goto fail_exit;
-  }
-
-  if (pid == 0) { // Child
-    // Restore old sigmask.
-    if (sigprocmask(SIG_SETMASK, &oldmask, NULL) == -1) {
-      std::cerr << cmd << ": sigprocmask failed: " << cpp_strerror(errno) << "\n";
-      goto fail_exit;
-    }
-    (void)setpgid(0, 0); // Become process group leader.
-    SubProcess::exec();
-    ceph_abort(); // Never reached
-  }
-
-  // Parent
-  (void)alarm(timeout);
-
-  for (;;) {
-    int signo;
-    if (sigwait(&mask, &signo) == -1) {
-      std::cerr << cmd << ": sigwait failed: " << cpp_strerror(errno) << "\n";
-      goto fail_exit;
-    }
-    switch (signo) {
-    case SIGCHLD:
-      int status;
-      if (waitpid(pid, &status, WNOHANG) == -1) {
-	std::cerr << cmd << ": waitpid failed: " << cpp_strerror(errno) << "\n";
-	goto fail_exit;
-      }
-      if (WIFEXITED(status))
-	_exit(WEXITSTATUS(status));
-      if (WIFSIGNALED(status))
-	_exit(128 + WTERMSIG(status));
-      std::cerr << cmd << ": unknown status returned\n";
-      goto fail_exit;
-    case SIGINT:
-    case SIGTERM:
-      // Pass SIGINT and SIGTERM, which are usually used to terminate
-      // a process, to the child.
-      if (::kill(pid, signo) == -1) {
-	std::cerr << cmd << ": kill failed: " << cpp_strerror(errno) << "\n";
-	goto fail_exit;
-      }
-      continue;
-    case SIGALRM:
-      std::cerr << cmd << ": timed out (" << timeout << " sec)\n";
-      if (::killpg(pid, sigkill) == -1) {
-	std::cerr << cmd << ": kill failed: " << cpp_strerror(errno) << "\n";
-	goto fail_exit;
-      }
-      continue;
-    default:
-      std::cerr << cmd << ": sigwait: invalid signal: " << signo << "\n";
-      goto fail_exit;
-    }
-  }
-
-fail_exit:
-  _exit(EXIT_FAILURE);
-}
+static void timeout_sighandler(int);
+static void dummy_sighandler(int);
 
 #endif
+#pragma once
+#include "SubProcess.cc"
```

Comment:
What is the point of moving things into a .cc file only to include the .cc file?

Reply:
@adamemerson point is to separate definitions from declarations

Reply:
@shinobu-x 
There are quite a lot of files where the inline definitions are in the .h file.
You do not want to change all of them, i think? So why do it here??

Next to the fact that it is very, very weird to include a `.cc` file in a `.h` file.
The/My idea is that the `.h` file is more or leas the stable part of the code, and serves as a API for that code block. And modifications in the code only effect that one source modules. And eases compilation. 
That last part is sort of complicated by the fact that there are a lot of inline functions, which need to go into the header file. And here the question is valid if all these functions need to be inlined? 
It looks like quite a few are have calls to other functions that might not be inlined, or that do have a rather long runtime, or are called very infrequently.... All reason not to inline functions.

So I guess if you want to fix this module, focus on what needs to be inlined, and what can actually do without. And put the later ones in `.cc` and have the function header in `.h`